src/language/rule/rule.ts

```diff
@@ -265,3 +278,37 @@ export class RuleFailure {
         return new RuleFailurePosition(position, lineAndCharacter);
     }
 }
+
+export class WalkContext<T> {
+    private limit: number;
+    private failures: RuleFailure[];
+
+    constructor(public readonly sourceFile: ts.SourceFile, public readonly ruleName: string, public readonly options: T) {
+        this.failures = [];
+        this.limit = sourceFile.getFullWidth();
+    }
+
+    public getFailures() {
+        return this.failures;
+    }
+
+    /** Add a failure with any arbitrary span. Prefer `addFailureAtNode` if possible. */
+    public addFailureAt(start: number, width: number, failure: string, fix?: Fix) {
+        this.addFailure(start, start + width, failure, fix);
+    }
+
+    public addFailure(start: number, end: number, failure: string, fix?: Fix) {
```

Comment:
I would prefer to leave it as it is now. This new implementation does not need to keep the mistakes made in `RuleWalker`.
When migrating an existing rule, you will have *some* conflicts anyway because of the other missing methods